flood-for-transmission: init at 2023-11-17T12-46-13

al3xtjames opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This change packages [flood-for-transmission](https://github.com/johman10/flood-for-transmission), an alternative web UI for Transmission. It's a Transmission-specific clone of [Flood](https://github.com/jesec/flood). The `services.transmission.webHome` option has also been added. It can be used to override the original web UI:

```nix
{ pkgs, ... }:

{
  services.transmission.webHome = pkgs.flood-for-transmission;
}
```

The service takes care of adding the path in `services.transmission.webHome` to the AppArmor configuration, so this works without having to override `extraAppArmorPaths` in `pkgs.transmission` or `pkgs.transmission_4`

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

al3xtjames committed:
flood-for-transmission: init at 2023-11-17T12-46-13

evanrichter commented on pkgs/applications/networking/p2p/flood-for-transmission/default.nix:
```diff
@@ -0,0 +1,34 @@
+  pname = "flood-for-transmission";
+  version = "2023-04-18T18-03-53";
+
+  src = fetchzip {
```
I'm new to nixpkgs maintenance, but I see more commonly fetchFromGithub:

```nix
  src = fetchFromGitHub {
    owner = "johnam10";
    repo = pname;
    rev = version;
    sha256 = "sha256-alsHOTF8EEF7iUNHvLC21V3VRVoYQSs78g2r7YGTDeQ=";
  };
```

al3xtjames replied:
I'm using `fetchzip` to download the compiled release. My understanding is that `fetchfromGitHub` fetches the source tarball.

evanrichter replied:
Ah so that is a significant difference then! Building from sources is preferred whenever possible. Imagine a problem discovered with nodejs produced frontends that makes a security issue in some packages. If you apply the patch to nodejs, it will rebuild all frontend packages and fix the problem for all.

Note that when this is merged in nixpkgs, hydra will build from source and offer users the built binary from cache.nixos.org

al3xtjames replied:
Thanks for the suggestion. I was unfamiliar with the Node packaging infrastructure in Nix, but I managed to update the derivation to use `buildNpmPackage` to build the package from source.

ambroisie commented:
I can use `flood` just fine with `transmission` in [my config](https://git.belanyi.fr/ambroisie/nix-config/src/branch/main/modules/services/flood/default.nix). Is it just so you don't have to run a different service?

al3xtjames commented:
Yeah, the advantage of `flood-for-transmission` is that it replaces the existing web interface (without requiring another service).

I'm using the hardened profile (`imports = [ (modulesPath + "/profiles/hardened.nix") ];`) which [enables AppArmor by default](https://github.com/NixOS/nixpkgs/blob/420250840fc7a672563c3242b95d30e116d30146/nixos/modules/profiles/hardened.nix#L38). The AppArmor profile for `transmission-daemon` in nixpkgs [uses an allowlist to limit the files it can access](https://github.com/NixOS/nixpkgs/blob/66aedfd010204949cb225cf749be08cb13ce1813/pkgs/applications/networking/p2p/transmission/default.nix#L118), including the default `TRANSMISSION_WEB_HOME`. This breaks if you you override `TRANSMISSION_WEB_HOME`, which is why I'm using an overlay instead of just setting `systemd.services.transmission.environment.TRANSMISSION_WEB_HOME` to `pkgs.flood-for-transmission`. I don't think there's an easy way to override the AppArmor profile (and I think it would require a rebuild anyways).

doronbehar commented:
@al3xtjames the package seems great, but I'm missing documentation for how to use this package. What if you'd add an additional commit with a new option to the `nixos/transmission` service, that adds support for this web UI via `TRANSMISSION_WEB_HOME`? Perhaps in the documentation of that option, you could mention that for users with AppArmor enabled, it won't work, and one has to use an overlay.

al3xtjames committed:
nixos/transmission: add webHome option

al3xtjames committed:
maintainers: add al3xtjames

al3xtjames commented:
Thanks for the feedback. I added a `webHome` option to `services.transmission.webHome` which sets `TRANSMISSION_WEB_HOME` in the service's environment variables and updates the AppArmor rules as needed (no need for an override). This makes `extraAppArmorPaths` (added in https://github.com/NixOS/nixpkgs/pull/259663) redundant. Is it worth reverting that change to `transmission_4`?

doronbehar commented on nixos/modules/services/torrent/transmission.nix:
```diff
@@ -251,6 +251,16 @@ in
+        default = null;
+        example = "pkgs.flood-for-transmission";
+        description = lib.mdDoc ''
+          Used as the value of `TRANSMISSION_WEB_HOME` if not `null`.
```
Here's a bit better and more informative documentation:

```suggestion
          If not `null`, sets the value of the `TRANSMISSION_WEB_HOME`
          environment variable used by the service. Useful for overriding
          the web interface files, without overriding the transmission
          package and thus requiring rebuilding it locally. Use this if
          you want to use the web interface provided by
          `pkgs.flood-for-transmission`.
```

Perhaps you could even modify the last sentence suggested above to write exactly how to use the package with this option? I mean, is it simply:

```nix
services.transmission.webHome = pkgs.flood-for-transmission;
```

Or is it:

```nix
services.transmission.webHome = "${pkgs.flood-for-transmission}/public";
```

Or something alike?

al3xtjames replied:
Thanks! I followed your suggestion for the most part. Regarding the usage of this option, the first example is enough.

doronbehar commented on nixos/modules/services/torrent/transmission.nix:
```diff
@@ -492,6 +503,10 @@ in
         px ${cfg.settings.script-torrent-done-filename} -> &@{dirs},
       ''}
+
+      ${optionalString (cfg.webHome != null) ''
+        r ${cfg.webHome}/**,
+      ''}
```
This is great. Regarding the `extraAppArmorPaths` option added in https://github.com/NixOS/nixpkgs/pull/259663 , I'm not familiar with using AppArmor, so I'll ask: In general with transmission, if these AppArmor paths are considered only when the daemon runs - in the NixOS service, perhaps we should add an `extraAppArmorPaths` option to the NixOS service instead of the package?

The guiding principle is that overriding a package is more expensive in terms of building resources in comparison to changing options in the service. If the above suggestion would be acceptable for you, I'd also revert 71ab0b6c1c94fdc5d702bbe401bbf841f0d2257b in the same commit.

al3xtjames replied:
I added `extraAppArmorPaths` to the derivation in order to support `flood-for-transmission` with AppArmor enabled. I think it's reasonable to revert it there; I'm not sure if it's worth porting it over to the module if there's no other use for it.

doronbehar replied:
> I added `extraAppArmorPaths` to the derivation in order to support `flood-for-transmission` with AppArmor enabled. I think it's reasonable to revert it there; I'm not sure if it's worth porting it over to the module if there's no other use for it.

I think that reverting that commit in this PR, in a separate commit is most appropriate - since the `webHome` committed here is a better solution to the AppArmor issue you fixed there.

I also think that there could be other usages for additional AppArmor paths, but that could be left to another PR or for another user to find the need for it.

al3xtjames replied:
Sounds reasonable, I reverted it in a separate commit.

doronbehar commented on pkgs/applications/networking/p2p/flood-for-transmission/default.nix:
```diff
@@ -0,0 +1,36 @@
+{ lib
```
In the 3rd commit, add this documentation please:

```suggestion
# To use this package, use: `services.transmission.webHome = ${pkgs.flood-for-transmission}/...`
{ lib
```

doronbehar reviewed (commented):
Hey, thanks for cooperating! This is getting in shape. A few improvements suggestions.

al3xtjames committed:
transmission_4: remove extraAppArmorPaths

extraAppArmorPaths was added in https://github.com/NixOS/nixpkgs/pull/259663
to extend the default AppArmor ruleset for transmission-daemon. This was
intended to support alternative web interfaces, which is no longer
needed with https://github.com/NixOS/nixpkgs/pull/241247. The NixOS
service was updated to add a user-specified TRANSMISSION_WEB_HOME to the
AppArmor profile without an override.

doronbehar commented on pkgs/applications/networking/p2p/flood-for-transmission/default.nix:
```diff
@@ -0,0 +1,37 @@
+# To use this package, use: `services.transmission.webHome = pkgs.flood-for-transmission};`
```
Nit:

```suggestion
# To use this package, use: `services.transmission.webHome = pkgs.flood-for-transmission;`
```

doronbehar reviewed (commented):
Seems great besides that typo! Will merge when CI is green.

doronbehar commented:
For the record, the Darwin build is failing, and this is unrelated to the change here, as already mentioned here: https://github.com/NixOS/nixpkgs/pull/272837 . Merging.

doronbehar commented:
Thanks for the contribution!

doronbehar commented:
Welcome to the Nixpkgs team.

al3xtjames commented:
Thanks!

dpatterbee commented:
I think this change broke the transmission module when `webHome` is not set because we now set `TRANSMISSION_WEB_HOME=` in that case. According to [the docs](https://github.com/transmission/transmission/blob/646883174b2a589834e1e70838de66df0ccbcba4/docs/Environment-Variables.md?plain=1#L11), transmission will use that env var if it is set - not if it is nonempty. It broke my transmission install at least and setting `webHome` fixed it.

I think the solution here is to set a default for the path, something like `pkgs.transmission + /share/transmission/web`.

Let me know if I should raise an issue for this.

@al3xtjames @doronbehar

al3xtjames commented:
I apologize for the regression. I though I had tested this change without setting `webHome` but it's clear that my testing was insufficient. I believe https://github.com/NixOS/nixpkgs/pull/278975 should fix it.

dpatterbee commented:
Ah I hadn't seen the PR for the fix, looks like everything's under control. I guess that's why it's called unstable :) the system works. Thanks
